src/Authentication/Authentication/Common/GraphSettings.cs

```diff
@@ -0,0 +1,312 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph.PowerShell.Authentication
+{
+    using Microsoft.Graph.PowerShell.Authentication.Common;
+    using Microsoft.Graph.PowerShell.Authentication.Extensions;
+    using Microsoft.Graph.PowerShell.Authentication.Helpers;
+    using Newtonsoft.Json;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Xml.Serialization;
+
+    /// <summary>
+    /// Represents a Microsoft Graph settings structure with environments.
+    /// </summary>
+    [Serializable]
+    internal class GraphSettings
+    {
+        public IDictionary<string, IGraphEnvironment> EnvironmentTable { get; set; } = new ConcurrentDictionary<string, IGraphEnvironment>(StringComparer.CurrentCultureIgnoreCase);
+
+        [JsonIgnore]
+        [XmlIgnore]
+        public IEnumerable<IGraphEnvironment> Environments
+        {
+            get
+            {
+                return EnvironmentTable.Values.ToList();
+            }
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="GraphSettings"/>.
+        /// </summary>
+        public GraphSettings()
+        {
+            EnvironmentTable = new ConcurrentDictionary<string, IGraphEnvironment>(StringComparer.CurrentCultureIgnoreCase);
+
+            foreach (var env in GraphEnvironment.BuiltInEnvironments)
```

Comment:
we seem to copy these builtinenvironments around quite a bit.  GraphEnvironmentConstants to GraphEnvironment to GraphSettings

Reply:
not sure my complaint here though so disregard.